content/en/docs/ops/diagnostic-tools/istioctl-analyze/test.sh

```diff
@@ -0,0 +1,115 @@
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e  # Exit on failure
+set -u  # Unset is an error
+# There is no need to echo, output appears in ops_diagnostic-tools_istioctl-analyze_test_debug.txt
+set -o pipefail
+
+# This script doesn't need a control plane initially and will install Istio when needed
+# @setup profile=none
+
+# The test harness labels the default namespace.  Remove that label
+# so the output matches the expect output on a fresh K8s cluster.
+kubectl label namespace default istio-injection- || true
+
+echo '*** istioctl-analyze step 1 ***'
+_verify_contains snip_analyze_all_namespaces "$snip_analyze_all_namespace_sample_response"
+
+echo '*** istioctl-analyze step 2 ***'
+snip_fix_default_namespace
+_verify_same snip_try_with_fixed_namespace "$snip_try_with_fixed_namespace_out"
+
+echo '*** istioctl-analyze step 3 ***'
+_verify_contains snip_analyze_sample_destrule "$snip_analyze_sample_destrule_out"
+
+# TODO It is unclear if this should ALWAYS fail, NEVER fail, or sometimes fail.
+# There are multiple DestinationRules, some are valid for the VirtualService, some lack subsets 
+echo '*** istioctl-analyze step ***'
+snip_analyze_networking_directory || true
+
+echo '*** istioctl-analyze step 5 ***'
+snip_analyze_all_networking_yaml
+
+echo '*** istioctl-analyze step 6 ***'
+snip_analyze_all_networking_yaml_no_kube
+
+echo '*** istioctl-analyze step 7 ***'
+istioctl analyze --help
+
+echo '*** istioctl-analyze step 8 ***'
+snip_install_with_custom_config_analysis
+_wait_for_deployment istio-system istiod
+
+echo '*** istioctl-analyze step 9 ***'
+set +e  # Don't exit on failure
+kubectl apply -f - <<EOF
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: ratings
+  namespace: default
+spec:
+  gateways:
+  - bogus-gateway
+  hosts:
+  - ratings
+  http:
+  - route:
+    - destination:
+        host: ratings
+EOF
+set -e  # Exit on failure
+
+echo '*** istioctl-analyze step 10 ***'
+get_ratings_virtual_service() {
+kubectl get vs ratings -o yaml
+}
+_verify_elided get_ratings_virtual_service "$snip_vs_yaml_with_status"
+
+echo '*** istioctl-analyze step 11 ***'
+kubectl create ns frod
+_verify_failure analyze_k_frod
+# TODO Replace the above with _verify_failure_same when it is available
+# _verify_same analyze_k_frod "$snip_analyze_k_frod_out"
+
+echo '*** istioctl-analyze step 12 ***'
+_verify_same snip_analyze_suppress0102 "$snip_analyze_suppress0102_out"
+
+echo '*** istioctl-analyze step 13 ***'
+# TODO This will return errors on non-Istio namespaces, e.g. ibm-cert-store and kube-node-lease
+# It is hard to know what to supply to fix that problem for testing.
```

Comment:
You could create a broken deployment in ns baz and use `_verify_contains` to check for something expected in the ouput. Then the next 2 snips could use `_verify_not_contains` to confirm the same string is not in the output any more. The request routing task does something similar in its test: https://github.com/istio/istio.io/blob/54113c6864f28d4eaee6b58154ecb85f7faa1789/content/en/docs/tasks/traffic-management/request-routing/test.sh#L72

Reply:
@frankbu I'd like to keep the user commands the same and use `_verify_not_contains`.  I tried it, and it works, but I suspect `not_contains` will suffer because changes that affect CLI output will not be detected.

Reply:
Yes, the safest way to use `_verify_not_contains` is in conjunction with `_verify_contains`, i.e., check that some output is there, and then after changing something, check that it's not. Another useful `_verify_` function is `_verify_lines` which can check that 0 or more lines appear in the output AND 0 or more lines do not.